### PR TITLE
Gallery - context menu - 2nd Attempt

### LINKED ISSUE #891: add context menu for gallery and documents
long-tapping a context menu in gallery or  documents should open a context menu in the same style as the one that appears when a chat-bubble is long-tapped.

the context-menu should offer at least "delete", share would also be nice (later one sth. as "jump to message", however, that requires probably #888 to be ready and is out of scope for now)

### LINKED ISSUE #944: add option to delete image from gallery
currently, there is no option to delete an image from "profile / images and videos".

enlarging the image allows swiping and sharing, but no delete button - afaik, this enlarged view is not really under our control, so it is probably hard to get a button in here.

therefore i suggest to show a simple "Delete" button on a long-tap in the overview - just the same as we are using in the chatview.

tappign the button should do the same as the "delete" button in the chat does - let the user confirm via safeActionSheet and call deleteMessage(). refreshing should be done automatically by listening to the DC_EVENT_MSGS_CHANGED (that might already be implemented)

### PR DESCRIPTION
**Pull request by nayooti**
closes #944 
closes #891 

Since MessageKit-files were deleted I was not able to rebase properly, so I had to rebase "manually" using this branch.

## This PR:
- context menu now displays (and plays) videos like on Telegram.
- context menu works on different device rotations 
- context menu closes on device rotation -> this is intended native behaviour
- context menu for documents gallery in iOS 13
- `contextMenuProvider` to generate legacy and new styled context menus 
- context menu provides **show in chat** that redirects to the message in chat view
- same menu applied to `DocumentGalleryController` (modern context menu uses default preview)
 
Gallery did not look nice in landscape on devices with notch. I did a quick fix in this PR too.

## To test: 
- use videos+images in landscape and portrait 
- try device in portrait and landscape

replaces #952

### CODE REVIEW

**Comment by cyBerta** [comment]
I just had a short check on my phone. It looks great, video works too. What is missing is support for gif files. At least the first image of an gif should be shown, if previewing gifs is not possible. Currently there's a white image shown.

**Review by cyBerta — CHANGES_REQUESTED** [review]
see my comment

File: deltachat-ios/Chat/ChatViewController.swift [hunk 1]
```diff
@@ -260,6 +259,7 @@ class ChatViewController: UITableViewController {
         if RelayHelper.sharedInstance.isForwarding() {
             askToForwardMessage()
         }
+        prepareContextMenu()
```
Comment (r10s) [inline deltachat-ios/Chat/ChatViewController.swift:262]: how is this change related to the gallery context menu? or what does that fix?

Comment (nayooti) [reply deltachat-ios/Chat/ChatViewController.swift:262]: `UIMenuController` is a singleton (by Apple). I change it's options in Gallery, so I have to reset these options when moving back from Gallery.

**Comment by r10s** [comment]
great!

some ui things:
- the delete-button should be red (see other apps)
- once tapped, there should be a confirmation (also see other apps)

to the general understanding: 
- the menu is a normal ViewController, or? os maybe call it ContextMenuViewController.swift to point out that it is just this. i was a bit confused what a MenuController is :)
- all in all this is some code and complexity - if we want to get the menu to the document view and - even more interestingly - to the normal chatview, we can resuse ContextMenuViewController.swift? (of course, we would need to add renderings for normal messages and files then)

File: deltachat-ios/Chat/ChatViewController.swift (hunk 1 above)
Comment (r10s) [reply deltachat-ios/Chat/ChatViewController.swift:262]: thanks for explanation :)

**Comment by nayooti** [comment]
`MenuController` is from Apple to show the old context menu (iOS 12 and lower)
`ContextMenuViewController' to show new context menu (iOS 13 and after).

`ContextMenuViewController' can be easily be reused and extended. For example to be used in `ChatViewController`, but obviously this needs to be extended for different MessageTypes. 

The complexity is only that we have to cover both context-menu-styles as long we support iOS12 and lower.
After all, the complexity isn't too bad IMO.

**Comment by nayooti** [comment]
@r10s I could not find any suitable confirm-delete-string. Any idea?

**Comment by r10s** [comment]
we're getting closer :)

some more things, **for ios 11, 12:**

- when the old-style context menu is opened, a tap outside the context menu should just close it. however, then this tap goes to the image, it also enlarges the image. this enlargement should require a second tap.

you can test this eg. in "iPhone 8 (iOS 12.2)" emulator

things **for ios 13+**:

- when the new-style context menu is opened, a tap outside the context menu should close it. this is correct. a tap _inside, on the preview image_ should directly enlarge the image. please check the normal ios gallery to get the idea.

- the delete confirmation is okayish, however, the ios gallery stays in the context-menu for the confirmation. this is no blocker, however, if this is easily doable by only setting a flag or so, we should just do it the same way. if it adds complexity, however, we should postpone it.

**Comment by nayooti** [comment]
I looked into this, and I came to the conclustion, that none of these "things" seem to be makable without much effort. 
Options are very limited. 

For iOS 13-context menu:
I tried adding gesture recognisers to the preview-view, but all touches were simply ignored. The preview-view closes no matter what.
Out of interest I added a button to that view, and also here I could not get it's action triggered.

In general I think Apple catches all the touches before they hit the view/menus. I assume they heavily use `hitTestOnEvent` here, where you can decide if you let touches pass though or not. 

The only idea I have right now is to override `hitTest`-method by ourself, but this is never beautiful as it will implemented on top of the whole app. 
 ... and it is not guaranteed to work (because I think Apple already is overriding/using it here)

**Comment by nayooti** [comment]
@r10s Ok, I am done for now.

File: deltachat-ios/Chat/ChatViewController.swift [hunk 2]
```diff
@@ -162,7 +162,13 @@ class ChatViewController: UITableViewController {
                                        name: UIApplication.willResignActiveNotification,
                                        object: nil)
         notificationCenter.addObserver(self, selector: #selector(keyboardWillShow(_:)), name: UIResponder.keyboardWillShowNotification, object: nil)
+<<<<<<< HEAD
```
Comment (cyBerta) [inline deltachat-ios/Chat/ChatViewController.swift:165]: That won't compile, I think.

**Review by cyBerta — CHANGES_REQUESTED** [review]

**Comment by cyBerta** [comment]
There's an unresolved merge conflict and unfortunately also new merge conflcts with latest master that need to be resolved, too.

**Review by cyBerta — APPROVED** [review]
lgtm, I fixed the merge conflict and removed the controls from the video preview, because a tap on any item within the preview will open the preview controller, so the controls can never be triggered. It's now the same UI as in the iOS Gallery wrt. video previews.
